src/tests/profiler/gc/nongcheap.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Profiler.Tests
+{
+    class NonGCHeapTests
+    {
+        static readonly Guid GcAllocateEventsProfilerGuid = new Guid("EF0D191C-3FC7-4311-88AF-E474CBEB2859");
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void AllocateNonGcHeapObjects()
+        {
+            // When this method is invoked, JIT is expected to trigger allocations for these
+            // string literals and they're expected to end up in a nongc segment (also known as frozen)
+            Consume("string1");
+            Consume("string2");
+            Consume("string3");
+            Consume("string4");
+            Consume("string5");
+            Consume("string6");
+
+            int gen = GC.GetGeneration("string7");
+            if (gen != int.MaxValue)
+                throw new Exception("object is expected to be in a non-gc heaps");
+
+            GC.Collect(gen);
```

Comment:
The current docs for GC.Collect() and GC.MaxGeneration [states](https://learn.microsoft.com/en-us/dotnet/api/system.gc.maxgeneration?view=net-8.0#remarks):

> Use the [MaxGeneration](https://learn.microsoft.com/en-us/dotnet/api/system.gc.maxgeneration?view=net-8.0#system-gc-maxgeneration) property to determine the maximum value you can specify when calling the [Collect(Int32)](https://learn.microsoft.com/en-us/dotnet/api/system.gc.collect?view=net-8.0#system-gc-collect(system-int32)) method that takes a generation parameter.

I don't think we need to promise anyone that GC.Collect(int.MaxValue) will work even if it happens to do so in the current implementation. Instead I think we would document something like:

> When GC.GetGeneration() is called with an object that doesn't exist on the GC heap, these objects do not have any associated GC generation. GetGeneration() will return int.MaxValue in this case and this value should not be treated as a valid GC generation for any other GC API that needs a generation specified.

Reply:
We should only call GC.Collect(Int32.MaxValue) in the test if we consider it a contract and are willing to fix it if it breaks. If we decide to not view that as a contract we shouldn't have it in the test

Reply:
`GC.Collect(Int32.MaxValue)` always worked fine. Changing it would be a breaking change.

I do not think the test coverage for it belongs to this test. A better place to add a test coverage for it would be https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/tests/System/GCTests.cs .

Reply:
Move to GCTests.cs, thanks!

Reply:
> Changing it would be a breaking change

In case there was any confusion I was not proposing we change GC.Collect() behavior. I wanted to avoid declaring in our docs that GC.Collect(int.MaxValue) was supported and well-defined.